File: deck/R/generate_tables.py
#!/usr/bin/env python3
"""
generate_tables.py
Generates all tables for the Data Science Introduction deck
Outputs: TeX files for direct inclusion in Beamer
"""

import os

# Set output directory
output_dir = "../tables"
os.makedirs(output_dir, exist_ok=True)

# =============================================================================
# Table 1: Data Types Overview
# =============================================================================
print("Generating Table 1: Data types...")

table1 = r"""\begin{tabular}{@{}llll@{}}
\toprule
\textbf{Type} & \textbf{Description} & \textbf{Examples} & \textbf{Share} \\
\midrule
Structured & Organized in tables/rows & SQL databases, CSV & 20\% \\
Semi-structured & Tagged or marked up & JSON, XML, HTML & 10\% \\
Unstructured & No predefined format & Text, images, video & 70\% \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_data_types.tex"), "w") as f:
    f.write(table1)

# =============================================================================
# Table 2: Machine Learning Algorithms Comparison
# =============================================================================
print("Generating Table 2: ML algorithms comparison...")

table2 = r"""\begin{tabular}{@{}llcc@{}}
\toprule
\textbf{Algorithm} & \textbf{Task} & \textbf{Interpret.} & \textbf{Scale} \\
\midrule
Linear Regression & Regression & High & Excellent \\
Logistic Regression & Classification & High & Excellent \\
Decision Trees & Both & Medium & Good \\
Random Forest & Both & Low & Good \\
Neural Networks & Both & Low & Varies \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_ml_algorithms.tex"), "w") as f:
    f.write(table2)

# =============================================================================
# Table 3: Programming Languages for Data Science
# =============================================================================
print("Generating Table 3: Programming languages...")

table3 = r"""\begin{tabular}{@{}lll@{}}
\toprule
\textbf{Language} & \textbf{Key Strengths} & \textbf{Rank} \\
\midrule
Python & General purpose, ML libraries & 1st \\
R & Statistics, visualization & 2nd \\
SQL & Data querying, databases & 3rd \\
Julia & High performance computing & Growing \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_languages.tex"), "w") as f:
    f.write(table3)

# =============================================================================
# Table 4: Key Metrics by Domain
# =============================================================================
print("Generating Table 4: Domain metrics...")

table4 = r"""\begin{tabular}{@{}lll@{}}
\toprule
\textbf{Domain} & \textbf{Key Metric} & \textbf{Challenge} \\
\midrule
Healthcare & Patient outcomes & Privacy (HIPAA) \\
Finance & ROI, Risk & Regulation \\
Marketing & Conversion rate & Attribution \\
Manufacturing & Defect rate & Real-time data \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_domains.tex"), "w") as f:
    f.write(table4)

# =============================================================================
# Table 5: Course Roadmap
# =============================================================================
print("Generating Table 5: Course roadmap...")

table5 = r"""\begin{tabular}{@{}cll@{}}
\toprule
\textbf{Weeks} & \textbf{Topic} & \textbf{Key Skills} \\
\midrule
1--3 & Foundations & Python, Statistics basics \\
4--6 & Data Wrangling & pandas, data cleaning \\
7--9 & Machine Learning & scikit-learn \\
10--12 & Deep Learning & PyTorch basics \\
13--15 & Capstone Project & Full pipeline \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_roadmap.tex"), "w") as f:
    f.write(table5)

# =============================================================================
# Table 6: Evaluation Metrics Summary
# =============================================================================
print("Generating Table 6: Evaluation metrics...")

table6 = r"""\begin{tabular}{@{}lll@{}}
\toprule
\textbf{Metric} & \textbf{Formula} & \textbf{Best For} \\
\midrule
Accuracy & $\frac{TP+TN}{\text{Total}}$ & Balanced data \\
Precision & $\frac{TP}{TP+FP}$ & Cost of FP high \\
Recall & $\frac{TP}{TP+FN}$ & Cost of FN high \\
F1-Score & $\frac{2 \cdot P \cdot R}{P+R}$ & Imbalanced data \\
AUC-ROC & Area under curve & Threshold tuning \\
\bottomrule
\end{tabular}"""

with open(os.path.join(output_dir, "table_metrics.tex"), "w") as f:
    f.write(table6)

print("\n========================================")
print("All tables generated successfully!")
print(f"Output directory: {os.path.abspath(output_dir)}")
print("========================================")

File: deck/R/generate_tables.R
#!/usr/bin/env Rscript
# generate_tables.R
# Generates all tables for the Data Science Introduction deck
# Outputs: TeX files for direct inclusion in Beamer

# Load required packages
required_packages <- c("xtable", "knitr", "kableExtra")

for (pkg in required_packages) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    install.packages(pkg, repos = "https://cloud.r-project.org/")
    library(pkg, character.only = TRUE)
  }
}

# Set output directory
output_dir <- "../tables"
if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)

# =============================================================================
# Table 1: Data Types Overview
# =============================================================================
cat("Generating Table 1: Data types...\n")

data_types <- data.frame(
  Type = c("Structured", "Semi-structured", "Unstructured"),
  Description = c("Organized in tables/rows", "Tagged or marked up", "No predefined format"),
  Examples = c("SQL databases, CSV", "JSON, XML, HTML", "Text, images, video"),
  Percentage = c("20\\%", "10\\%", "70\\%")
)

# Generate LaTeX table
sink(file.path(output_dir, "table_data_types.tex"))
cat("\\begin{tabular}{@{}llll@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Type} & \\textbf{Description} & \\textbf{Examples} & \\textbf{Share} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(data_types)) {
  cat(paste(data_types$Type[i], "&",
            data_types$Description[i], "&",
            data_types$Examples[i], "&",
            data_types$Percentage[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

# =============================================================================
# Table 2: Machine Learning Algorithms Comparison
# =============================================================================
cat("Generating Table 2: ML algorithms comparison...\n")

ml_algorithms <- data.frame(
  Algorithm = c("Linear Reg.", "Logistic Reg.", "Decision Trees", "Random Forest", "Neural Nets"),
  Task = c("Regression", "Classification", "Both", "Both", "Both"),
  Interpretable = c("High", "High", "Medium", "Low", "Low"),
  Scalability = c("Excellent", "Excellent", "Good", "Good", "Varies")
)

sink(file.path(output_dir, "table_ml_algorithms.tex"))
cat("\\begin{tabular}{@{}llcc@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Algorithm} & \\textbf{Task} & \\textbf{Interpret.} & \\textbf{Scale} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(ml_algorithms)) {
  cat(paste(ml_algorithms$Algorithm[i], "&",
            ml_algorithms$Task[i], "&",
            ml_algorithms$Interpretable[i], "&",
            ml_algorithms$Scalability[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

# =============================================================================
# Table 3: Programming Languages for Data Science
# =============================================================================
cat("Generating Table 3: Programming languages...\n")

languages <- data.frame(
  Language = c("Python", "R", "SQL", "Julia"),
  Strengths = c("General purpose, ML", "Statistics, visualization", "Data querying", "High performance"),
  Adoption = c("1st", "2nd", "3rd", "Growing")
)

sink(file.path(output_dir, "table_languages.tex"))
cat("\\begin{tabular}{@{}lll@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Language} & \\textbf{Key Strengths} & \\textbf{Rank} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(languages)) {
  cat(paste(languages$Language[i], "&",
            languages$Strengths[i], "&",
            languages$Adoption[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

# =============================================================================
# Table 4: Key Metrics by Domain
# =============================================================================
cat("Generating Table 4: Domain metrics...\n")

domains <- data.frame(
  Domain = c("Healthcare", "Finance", "Marketing", "Manufacturing"),
  Metric = c("Patient outcomes", "ROI, Risk", "Conversion rate", "Defect rate"),
  Challenge = c("Privacy (HIPAA)", "Regulation", "Attribution", "Real-time data")
)

sink(file.path(output_dir, "table_domains.tex"))
cat("\\begin{tabular}{@{}lll@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Domain} & \\textbf{Key Metric} & \\textbf{Challenge} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(domains)) {
  cat(paste(domains$Domain[i], "&",
            domains$Metric[i], "&",
            domains$Challenge[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

# =============================================================================
# Table 5: Course Roadmap
# =============================================================================
cat("Generating Table 5: Course roadmap...\n")

roadmap <- data.frame(
  Week = c("1--3", "4--6", "7--9", "10--12", "13--15"),
  Topic = c("Foundations", "Data Wrangling", "Machine Learning", "Deep Learning", "Capstone"),
  Skills = c("Python, Stats basics", "pandas, cleaning", "scikit-learn", "PyTorch basics", "Full pipeline")
)

sink(file.path(output_dir, "table_roadmap.tex"))
cat("\\begin{tabular}{@{}cll@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Weeks} & \\textbf{Topic} & \\textbf{Key Skills} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(roadmap)) {
  cat(paste(roadmap$Week[i], "&",
            roadmap$Topic[i], "&",
            roadmap$Skills[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

# =============================================================================
# Table 6: Evaluation Metrics Summary
# =============================================================================
cat("Generating Table 6: Evaluation metrics...\n")

metrics <- data.frame(
  Metric = c("Accuracy", "Precision", "Recall", "F1-Score", "AUC-ROC"),
  Formula = c("$\\frac{TP+TN}{Total}$", "$\\frac{TP}{TP+FP}$",
              "$\\frac{TP}{TP+FN}$", "$\\frac{2 \\cdot P \\cdot R}{P+R}$", "Area under curve"),
  UseCase = c("Balanced data", "Cost of FP high", "Cost of FN high", "Imbalanced data", "Threshold tuning")
)

sink(file.path(output_dir, "table_metrics.tex"))
cat("\\begin{tabular}{@{}lll@{}}\n")
cat("\\toprule\n")
cat("\\textbf{Metric} & \\textbf{Formula} & \\textbf{Best For} \\\\\n")
cat("\\midrule\n")
for (i in 1:nrow(metrics)) {
  cat(paste(metrics$Metric[i], "&",
            metrics$Formula[i], "&",
            metrics$UseCase[i], "\\\\\n"))
}
cat("\\bottomrule\n")
cat("\\end{tabular}\n")
sink()

cat("\n========================================\n")
cat("All tables generated successfully!\n")
cat("Output directory:", normalizePath(output_dir), "\n")
cat("========================================\n")
